File: graph/island_perimeter.py
from typing import List
from collections import deque


class Solution:
    # dfs
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        visit = set()

        def dfs(i, j):
            if i >= len(grid) or j >= len(grid[0]) or i < 0 or j < 0 or grid[i][j] == 0:
                return 1
            if (i, j) in visit:
                return 0
            visit.add((i, j))
            perim = dfs(i, j + 1)
            perim += dfs(i, j - 1)
            perim += dfs(i + 1, j)
            perim += dfs(i - 1, j)
            return perim

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]:
                    return dfs(i, j)

    # bfs
    def islandPerimeter2(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visitSet = set()
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        perims = [0]

        def bfs(row, col):
            q = deque()
            q.append((row, col))
            visitSet.add((row, col))

            while q:
                r, c = q.popleft()
                for d in directions:
                    dr, dc = r + d[0], c + d[1]
                    if (dr < 0 or dr == ROWS) or (dc < 0 or dc == COLS):
                        perims[0] += 1
                    elif grid[dr][dc] == 0:
                        perims[0] += 1
                    elif grid[dr][dc] and (dr, dc) not in visitSet:
                        visitSet.add((dr, dc))
                        q.append((dr, dc))

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] and (r, c) not in visitSet:
                    bfs(r, c)

        return perims[0]

File: graph/island_perimeter.java
public class island_perimeter {
    
}
class Solution {
    public int islandPerimeter1(int[][] grid) {
        Set<String> visit = new HashSet<>();
        int row = 0, col = 0;
        for(int r = 0; r < grid.length; r++){
            for(int c = 0; c < grid[0].length; c++){
                if(grid[r][c] == 1){
                    row = r;
                    col = c;
                    break;
                }
            }
        }
        return dfs(grid, visit, row, col);
    }

    private int dfs(int[][] grid, Set<String> visit, int i, int j) {
        if (i >= grid.length || j >= grid[0].length || i < 0 || j < 0 || grid[i][j] == 0) {
            return 1;
        }
        if (visit.contains(i + "," + j)) {
            return 0;
        }
        visit.add(i + "," + j);
        int perim = dfs(grid, visit, i, j + 1);
        perim += dfs(grid, visit, i, j - 1);
        perim += dfs(grid, visit, i + 1, j);
        perim += dfs(grid, visit, i - 1, j);
        return perim;
    }

    private Set<int[]> visited = new HashSet<>();
    private int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    private int perims;
    private int ROWS;
    private int COLS;

    public int islandPerimeter(int[][] grid){
        ROWS = grid.length;
        COLS = grid[0].length;
        for(int r = 0; r < ROWS; r++){
            for(int c = 0; c < COLS; c++){
                if(grid[r][c] == 1 && !visited.contains(new int[]{r, c})){
                    bfs(r, c, grid);
                }
            }
        }

        return perims;

    }

    private void bfs(int row, int col, int[][] grid){
        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{row, col});
        visited.add(new int[]{row, col});

        while(!q.isEmpty()){
            int[] cell = q.poll();
            int r = cell[0];
            int c = cell[1];
            for(int[] d : directions){
                int dr = r + d[0];
                int dc = c + d[1];
                if(dr < 0 || dr == ROWS || dc < 0 || dc == COLS){
                    perims++;
                }
                else if(grid[dr][dc] == 0){
                    perims++;
                }
                else if(grid[dr][dc] == 1 && !visited.contains(new int[]{dr, dc})){
                    visited.add(new int[]{dr, dc});
                    q.offer(new int[]{dr, dc});
                }
            }
        }
    }
}
